### PR TITLE
Feature/152-Improve-documentation

### LINKED ISSUE #152: Improve documentation
Make sure purpose, main concepts, components, and installation/usage are clear.

Fix typos in intro paragraph and differentiate between main concepts and main dialogue components.

Ontology is now called domain.

Add ReadTheDocs badge to README.

### PR DESCRIPTION
**Pull request by NoB0**
First attempt to reorganise the documentation.
Changes include:
- Reorganize main concepts and components
- Fix some typos and rephrasing
- Add badge to README
- Replace overview figure

I am not sure the subsection `Rasa as a component library` in nlu.rst should be kept, what do you think?

Fixes #152

### CODE REVIEW

File: docs/source/participant.rst [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+:py:mod:`dialoguekit.participant.user`
+
+Represents a human interacting with an agent in natural language text.
+
+User preferences
+^^^^^^^^^^^^^^^^
+
+:py:mod:`dialoguekit.participant.user_preferences`
+
+User's preferences are expressed for specific slot-value pairs, where slots correspond to entities or properties in the domain.
```
Comment (kbalog) [inline docs/source/participant.rst:32]: Nit: "Represents the user's expressed preferences for specific..."

**Review by kbalog — APPROVED** [review]
LGTM

There are several minor issues, but those are mostly nits.
Let me know if anything needs further discussion.

File: docs/source/concepts.rst [hunk 2]
```diff
@@ -1,84 +1,42 @@
...
-Platform 
---------
-:py:mod:`dialoguekit.platforms.platform`
-
-The platform's responsibility is to facilitate the conversation. DialogueKit includes a simple terminal-based platform. However, it can support other platforms by facilitating communication via POST requests. 
 
-Dialogue Connector 
-----------------
-:py:mod:`dialoguekit.connector.dialogue_connector`
+A dialogue is turn-based and takes place between two participants (agent, user) that converse about a domain such as movies and flight booking.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/concepts.rst:20]: Done.

File: docs/source/concepts.rst [hunk 3]
```diff
@@ -1,84 +1,42 @@
...
+It is composed of a list of utterances that can be annotated or not and any relevant additional metadata.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/concepts.rst:21]: Done.

File: docs/source/concepts.rst [hunk 4]
```diff
@@ -1,84 +1,42 @@
...
+It is composed of a list of utterances that can be annotated or not and any relevant additional metadata. 
 
-Holds and orchestrates the conversation between the participants.
+Dialogue Orchestration
+----------------------
 
-Domain 
-------
-:py:mod:`dialoguekit.core.domain`
+The dialogue orchestration includes the dialogue connector and the platform that permit the participants to communicate.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/concepts.rst:26]: Done.

File: docs/source/concepts.rst [hunk 5]
```diff
@@ -1,84 +1,42 @@
...
 
-Defines the types of entities and the set of properties ("slots") for each entity type.
+Natural Language Understanding (NLU)
+------------------------------------
 
-Annotations
------------
-There are two types of annotations
+NLU is responsible for obtaining a structured representation of raw text utterances.
+Here, it entails intent classification, entity recognition, and user satisfaction prediction.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/concepts.rst:33]: Done.

File: docs/source/concepts.rst [hunk 6]
```diff
@@ -1,84 +1,42 @@
...
 
-* **Intent** :py:mod:`dialoguekit.core.intent`: represents the dialogue action.
-* **SlotValueAnnotation** :py:mod:`dialoguekit.core.slot_value_annotation`: slot-value pairs, where a slot refers to an entity or a property in the domain.
+For example in case of restaurant recommendation, the intent *DISCLOSE* and the slot value pair (*food*, *Turkish*) can be extracted from the utterance "I am looking for a cheap Turkish restaurant."
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/concepts.rst:35]: Done.

File: docs/source/concepts.rst [hunk 7]
```diff
@@ -1,84 +1,42 @@
...
 
+Natural Language Generation (NLG)
+---------------------------------
 
-User preferences
-----------------
+NLG is responsible for creating raw text utterances from their structured representation.
 
-* Preferences are expressed for specific slot-value pairs, where slots correspond to entities or properties in the domain.
+For example in the case of movie recommendation, "I recommend the movie Titanic." can be generated for the intent *RECOMMEND* and slot value pair (*title*, *Titanic*).
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/concepts.rst:42]: Done.

File: docs/source/core.rst [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+:py:mod:`dialoguekit.core.utterance`
+
+Dialogue participants exchange utterances, which are represented as raw text. 
+
+AnnotatedUtterance
+^^^^^^^^^^^^^^^^^^
+
+:py:mod:`dialoguekit.core.annotated_utterance`
+
+This class enrich Utterance with an intent, annotations, and/or freely definable metadata.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/core.rst:18]: Done.

File: docs/source/core.rst [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+:py:mod:`dialoguekit.core.domain`
+
+Defines the types of entities and the set of properties ("slots") for each entity type.
+
+Annotations
+-----------
+
+:py:mod:`dialoguekit.core.annotation`
+
+By default an annotation has a slot and a value, for example the slot *price* has the value *$10* in the following utterance *"This dish costs $10."*.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/core.rst:39]: Done.

File: docs/source/dialogue_orchestration.rst [hunk 10]
```diff
@@ -0,0 +1,21 @@
+Dialogue Orchestration
+======================
+
+DialogueKit use a connector to orchestrate the dialogue between the participants (i.e., agent and user).
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/dialogue_orchestration.rst:4]: Done.

File: docs/source/dialogue_orchestration.rst [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+Holds and orchestrates the conversation between the participants.
+
+Platform 
+--------
+
+:py:mod:`dialoguekit.platforms.platform`
+
+The platform's responsibility is to facilitate the conversation and ensure that the participant can see the agent's utterances and reply to it.
+DialogueKit includes a simple terminal-based platform.
+However, other platforms can be created by inheriting from :py:class:`dialoguekit.platforms.platform.Platform`.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/dialogue_orchestration.rst:21]: Good point, I also added `Flask chat` to the list of examples.

File: docs/source/index.rst [hunk 12]
```diff
@@ -1,7 +1,9 @@
 DialogueKit |release| documentation
-=====================================
+===================================
 
-DialogueKit is a framework for building and evaluating conversational information access systems.
+DialogueKit is a framework for building and evaluating conversational information access (CIA) systems.
+It contains based classes for fundamental concepts, such as dialogue participants, dialogue management, natural language understanding, and natural language generation.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/index.rst:5]: Done.

File: docs/source/index.rst [hunk 13]
```diff
@@ -1,7 +1,9 @@
...
+In addition to these concepts, DialogueKit contains an evaluation module to assess the performances of CIA systems at the dialogue level.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/index.rst:6]: Done.

File: docs/source/installation.rst [hunk 14]
```diff
@@ -4,9 +4,14 @@ Installation
 Install as a package
 --------------------
 
-**Note:** *Packaging is still a work in progress and may not work perfectly.*
+DialogueKit is published to PyPI, install it by running:
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/installation.rst:7]: Done.

File: docs/source/nlg.rst [hunk 15]
```diff
@@ -1,10 +1,12 @@
...
 
-The NLG components in Dialoguekit are currently template-based.
+The NLG components in DialogueKit are currently template-based.
 The basic template-based generation works by extracting templates from a set of training data. The templates are the user and agent utterances with the annotations removed. This version can be found here: :py:class:`dialoguekit.nlg.nlg_template.TemplateNLG`.
 
-An extended version of the basic template-based implementation which can perform conditional language generation is :py:class:`dialoguekit.nlg.nlg_conditional.ConditionalNLG`. Using a conditional may be useful in cases were other attributes then the intent and annotations may be of interest. Such as the users satisfaction. This can then be used to selecting templates that have a different tone based on the conditional, in this case the satisfaction score.
+An extended version of the basic template-based implementation which can perform conditional language generation is :py:class:`dialoguekit.nlg.nlg_conditional.ConditionalNLG`.
+Using a conditional may be useful in cases where other attributes than the intent and annotations may be of interest such as the users satisfaction.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/nlg.rst:8]: Done.

File: docs/source/nlu.rst [hunk 16]
```diff
@@ -8,36 +8,30 @@ Intent Classification
 
 Thus far two different NLU pipelines are implemented for intent classification
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/nlu.rst:9]: Done.

File: docs/source/nlu.rst [hunk 17]
```diff
@@ -8,36 +8,30 @@ Intent Classification
...
 
-The result
-""""""""""
+Implementation
+""""""""""""""
 
-The first rasa implementation is in *diet_classifier_rasa*, this model can be trained and thus uses multiple rasa components and structures.
-Below you can see all the imports that are used, only from rasa.
+The implementation is in *diet_classifier_rasa*, this model can be trained and thus uses multiple Rasa components and structures.
+Below you can see all the imports that are used, only from Rasa.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/nlu.rst:34]: Done.

File: docs/source/nlu.rst [hunk 18]
```diff
@@ -8,36 +8,30 @@ Intent Classification
...
+In general the idea was to import the necessary packages and re-implement the Rasa workflow with their components and structures. Rasa is built in a very object oriented structure. This does not allow us to use DialogueKit objects, they need to be transformed to Rasa components before use.
 
-The result
-""""""""""
+Implementation
+""""""""""""""
 
-The first rasa implementation is in *diet_classifier_rasa*, this model can be trained and thus uses multiple rasa components and structures.
-Below you can see all the imports that are used, only from rasa.
+The implementation is in *diet_classifier_rasa*, this model can be trained and thus uses multiple Rasa components and structures.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/nlu.rst:33]: Done.

File: docs/source/nlu.rst [hunk 19]
```diff
@@ -8,36 +8,30 @@ Intent Classification
...
+See below for an explanation of the implementation of Rasa DIET classifier.
 
 Rasa as a component library
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 *diet_classifier_rasa* implement Rasa's DIET classifier. This is a Dual Intent and Entity Transformer, their paper can be read 
 `here. <https://arxiv.org/pdf/2004.09936.pdf>`_
 
-Normally one would use Rasa as the underlying platform. But for our use-case we need to use it as a component in Dialoguekit. This proved to be possible with a bit of effort. Rasa is distributed with a Apache 2.0 license, granting us free use.
+Normally one would use Rasa as the underlying platform. But for our use-case we need to use it as a component in DialogueKit. Rasa is distributed with a Apache 2.0 license, granting us free use.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/nlu.rst:23]: Done.

File: docs/source/nlu.rst [hunk 20]
```diff
@@ -8,36 +8,30 @@ Intent Classification
...
 General idea
 """"""""""""
 
-In general the idea was to import the necessary packages and re-implement the rasa workflow with their components and structures. Rasa is built in a very object oriented structure. This does not allow us to use Dialoguekit objects, they need to be transformed to Rasa components before use.
-
-How we implemented it
-"""""""""""""""""""""
-
-As stated by a author on medium (link will be updated), the general structure of Rasa is a hard one to understand at first, but after a lot of trial and error we found the best way to get what we needed was to look at the implementation of the tests. They often show a minimal way of using the components separately and in conjunction with each other
+In general the idea was to import the necessary packages and re-implement the Rasa workflow with their components and structures. Rasa is built in a very object oriented structure. This does not allow us to use DialogueKit objects, they need to be transformed to Rasa components before use.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/nlu.rst:28]: Done.

File: docs/source/participant.rst [hunk 21]
```diff
@@ -0,0 +1,32 @@
+Participant 
+===========
+
+:py:mod:`dialoguekit.participant.participant`
+
+Agents and users are the participants in a dialogue. Generally, it is assumed that the agent is a conversational system and the user is a human.
+However, the agent might be played by a human ("Wizard-of-Oz") and the human user might be simulated.
+
+DialogueKit assumes the agent to always start the conversation and also end it. (A user can initialize ending a conversation, but the responsibility lies on the agent to actually stop it.)
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/participant.rst:9]: Done.

File: docs/source/participant.rst (hunk 1 above)
Comment (NoB0) [reply docs/source/participant.rst:32]: Done.

File: docs/source/external_agents.rst [hunk 22]
```diff
@@ -1,24 +1,31 @@
----
-orphan: true
----
-
 External Agents
 ===============
 
-Rasa parrot
---------------
+DialogueKit is shipped with 4 agents presented below.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/external_agents.rst:4]: Done.

File: docs/source/external_agents.rst [hunk 23]
```diff
@@ -1,24 +1,31 @@
...
+
+ParrotAgent
+-----------
+
+:py:mod:`sample_agents.parrot_agent`
+
+This agent will welcome the user, but will always parrot (echo) what the user says.
+
+RasaParrotAgent
+^^^^^^^^^^^^^^^
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/external_agents.rst:15]: Done.

File: docs/source/external_agents.rst [hunk 24]
```diff
@@ -1,24 +1,31 @@
...
 
-The :py:mod:`sample_agents.rasa_parrot_agent.py` **Agent** is just an example of how a Rasa agent could be used in DialogueKit.
-This **Agent** talks to the :py:mod:`additional.rasa-parrot` which does all the processing while the **DialogueKit Rasa Agent** only handles the communication.
+:py:mod:`sample_agents.rasa_parrot_agent`
 
+This agent is just an example of how a Rasa agent could be used in DialogueKit.
+It talks to the :py:mod:`additional.rasa-parrot` which does all the processing while the RasaParrotAgent only handles the communication.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/external_agents.rst:20]: Good catch, changed the name of the module.

File: docs/source/external_agents.rst [hunk 25]
```diff
@@ -62,6 +69,19 @@ An example can be seen below.
     dc.close()
 
 
-In this example we use a **User** to talk to the Rasa parroting service.
+In this example we use a user to talk to the Rasa parroting service.
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/external_agents.rst:72]: Done.

File: docs/source/core.rst [hunk 26]
```diff
@@ -0,0 +1,52 @@
+Core components
```
[the remark this answers is not in the document]
Comment (NoB0) [reply docs/source/core.rst]: Good point, I think "Dialogue Representation" is more intuitive. I changed the filename and title accordingly.

**Pull request merged**